### PR TITLE
Configurable retry logic samples

### PR DESCRIPTION
**Pull request by DavoudEshtehari**
New samples to address in ADO.NET documents.

### CODE REVIEW

File: doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml [hunk 1]
```diff
@@ -1064,6 +1064,9 @@ To apply the retry logic, do the following steps before opening the connection:
 > [NOTE!]
 > The connection **timeout** restarts for each execution of a connection open. There is no timing overlap between these two actions.
 
+> [NOTE!]
+> The default retry logic is none-retriable unless it was set in application configuration file. For more information see [Configurable retry logic and configuration file](/sql/connect/ado-net/configurable-retry-logic-config-file).
```
Comment (cheenamalhotra) [inline doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml:1068]:
```suggestion
> The default retry logic provider is not enabled unless it was configured in application configuration file. For more information see [Configurable retry logic and configuration file](/sql/connect/ado-net/configurable-retry-logic-config-file).
```

File: doc/snippets/Microsoft.Data.SqlClient/SqlCommand.xml [hunk 2]
```diff
@@ -2802,6 +2802,9 @@ To apply the retry logic, do the following steps before executing the command:
 > [NOTE!]
 > The command **timeout** restarts for each execution of a command within the retry logic and after applying the retry time delay. There is no timing overlap between these two actions.
 
+> [NOTE!]
+> The default retry logic is none-retriable unless it was set in application configuration file. For more information see [Configurable retry logic and configuration file](/sql/connect/ado-net/configurable-retry-logic-config-file).
```
Comment (cheenamalhotra) [inline doc/snippets/Microsoft.Data.SqlClient/SqlCommand.xml:2806]:
```suggestion
> The default retry logic provider is not enabled unless it was configured in application configuration file. For more information see [Configurable retry logic and configuration file](/sql/connect/ado-net/configurable-retry-logic-config-file).
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_OpenConnection.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+            NumberOfTries = 5,
+            // Prefered gap time to delay before retry
+            DeltaTime = TimeSpan.FromSeconds(1),
+            // Maximum gap time for each delay time before retry
+            MaxTimeInterval = TimeSpan.FromSeconds(20)
+        };
+        // </Snippet1>
+
+        // <Snippet2>
+        // Create a retry provider
```
Comment (cheenamalhotra) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_OpenConnection.cs:29]:
```suggestion
        // Create a retry logic provider
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_OpenConnection.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        SqlRetryLogicBaseProvider provider = SqlConfigurableRetryFactory.CreateExponentialRetryProvider(options);
+        // </Snippet2>
+
+        using(SqlConnection connection = new SqlConnection(connectionString))
+        {
+            // <Snippet3>
+            // Assumes that connection is a valid SqlConnection object 
+            // Assign a retry provider to the connection
```
Comment (cheenamalhotra) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_OpenConnection.cs:37]:
```suggestion
            // Set the retry logic provider on the connection instance
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+                DeltaTime = TimeSpan.FromSeconds(1),
+                // Maximum gap time for each delay time before retry
+                MaxTimeInterval = TimeSpan.FromSeconds(20),
+                // SqlException retriable error numbers
+                TransientErrors = new int[] { 4060, 1024, 1025}
+            };
+            // </Snippet1>
+
+            // <Snippet2>
+            // Create a custom retry provider
```
Comment (cheenamalhotra) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:38]:
```suggestion
            // Create a custom retry logic provider
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+            SqlRetryLogicBaseProvider provider = CustomRetry.CreateCustomProvider(options);
+            // </Snippet2>
+
+            using (SqlConnection connection = new SqlConnection(connectionString))
+            {
+                // <Snippet3>
+                // Assumes that connection is a valid SqlConnection object 
+                // Assign a retry provider to the connection
```
Comment (cheenamalhotra) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:46]:
```suggestion
                // Set the retry logic provider on the connection instance
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+                connection.RetryLogicProvider = provider;
+                // Establishing the connection will retry if one of the given transient failure or TimeoutException occurs.
```
Comment (cheenamalhotra) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:48]:
```suggestion
                // Establishing the connection will trigger retry if one of the given transient failure occurs.
```

I think you don't mean timeout here, plz check.

**Review by cheenamalhotra — APPROVED** [review]

File: doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml [hunk 8]
```diff
@@ -1058,12 +1058,15 @@ To apply the retry logic, do the following steps before opening the connection:
...
-> [NOTE!]
+> [!NOTE]
 > Detecting retriable exceptions is a vital part of the retry pattern. Before applying retry logic, it is important to investigate exceptions and choose a retry provider that best fits your scenario. First, log your exceptions and find transient faults.  
 
-> [NOTE!]
+> [!NOTE]
 > The connection **timeout** restarts for each execution of a connection open. There is no timing overlap between these two actions.
 
+> [!NOTE]
+> The default retry logic provider is not enabled unless it was configured in application configuration file. For more information see [Configurable retry logic and configuration file](/sql/connect/ado-net/configurable-retry-logic-config-file).
```
Comment (karinazhou) [inline doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml:1068]:
```suggestion
> The default retry logic provider is not enabled unless it was configured in application configuration file. For more information, see [Configurable retry logic and configuration file](/sql/connect/ado-net/configurable-retry-logic-config-file).
```

File: doc/snippets/Microsoft.Data.SqlClient/SqlCommand.xml [hunk 9]
```diff
@@ -2796,12 +2796,15 @@ To apply the retry logic, do the following steps before executing the command:
...
-> [NOTE!]
+> [!NOTE]
 > Detecting retriable exceptions is a vital part of the retry pattern. Before applying retry logic, it is important to investigate exceptions and choose a retry provider that best fits your scenario. First, log your exceptions and find transient faults.  
 
-> [NOTE!]
+> [!NOTE]
 > The command **timeout** restarts for each execution of a command within the retry logic and after applying the retry time delay. There is no timing overlap between these two actions.
 
+> [!NOTE]
+> The default retry logic provider is not enabled unless it was configured in application configuration file. For more information see [Configurable retry logic and configuration file](/sql/connect/ado-net/configurable-retry-logic-config-file).
```
Comment (karinazhou) [inline doc/snippets/Microsoft.Data.SqlClient/SqlCommand.xml:2806]:
```suggestion
> The default retry logic provider is not enabled unless it was configured in application configuration file. For more information, see [Configurable retry logic and configuration file](/sql/connect/ado-net/configurable-retry-logic-config-file).
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+
+        private static void RetryConnection(string connectionString)
+        {
+            // <Snippet1>
+            // Define the retry logic parameters
+            var options = new SqlRetryLogicOption()
+            {
+                // Tries 5 times before throwing an exception
+                NumberOfTries = 5,
+                // Prefered gap time to delay before retry
```
Comment (johnnypham) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:28]:
```suggestion
                // Preferred gap time to delay before retry
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+
+namespace CustomCRL_Doc
+{
+    class Program
+    {
+        private const string CnnStringFormat = "Server=localhost; Initial Catalog=Northwind__; Integrated Security=true; pooling=false; Timeout=1";
```
Comment (johnnypham) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:12]: Is the "__" prefix intentional?

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+                {
+                    // Check each error numbers if it was listed as a retriable error number
+                    if (retriableConditions.Contains(item.Number))
+                    {
+                        result = true;
+                        break;
+                    }
+                }
+            }
+            // Other type of exceptions can be assess separately
```
Comment (johnnypham) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:89]:
```suggestion
            // Other types of exceptions can be assessed separately
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+            }
+            return result;
+        }
+        // </Snippet5>
+    }
+
+    // <Snippet6>
+    public class CustomEnumerator : SqlRetryIntervalBaseEnumerator
+    {
+        // Set the maximum acceptable time to 4 minuts
```
Comment (johnnypham) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:102]:
```suggestion
        // Set the maximum acceptable time to 4 minutes
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+        private readonly TimeSpan _maxValue = TimeSpan.FromMinutes(4);
+
+        public CustomEnumerator(TimeSpan timeInterval, TimeSpan maxTime, TimeSpan minTime)
+            :base(timeInterval, maxTime, minTime) {}
```
Comment (johnnypham) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:106]:
```suggestion
            : base(timeInterval, maxTime, minTime) {}
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+            intervalTime = TimeSpan.Zero;
+            // First try has occurred before starting the retry process. 
+            // Check if it's still allowed to do retry
+            bool result = Current < NumberOfTries - 1;
+
+            if (result)
+            {
+                // Increase the number of attempts
+                Current++;
+                // It doesn't mind if the enumerator gets to the last value till the number of attempts ends.
```
Comment (johnnypham) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:181]:
```suggestion
                // It doesn't matter if the enumerator gets to the last value until the number of attempts end.
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+    {
+        // Preserve the given retryLogic on creation
+        public CustomProvider(SqlRetryLogicBase retryLogic)
+        {
+            RetryLogic = retryLogic;
+        }
+
+        public override TResult Execute<TResult>(object sender, Func<TResult> function)
+        {
+            // Create a list to save transient exceptions to report later if it was necessary
```
Comment (johnnypham) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:202]:
```suggestion
            // Create a list to save transient exceptions to report later if necessary
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+            IList<Exception> exceptions = new List<Exception>();
+            // Prepare it before reusing
+            RetryLogic.Reset();
+            // Create an infinite loop to keep attempt for defined maximum number of tries
```
Comment (johnnypham) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:206]:
```suggestion
            // Create an infinite loop to attempt the defined maximum number of tries
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_OpenConnection.cs [hunk 18]
```diff
@@ -0,0 +1,44 @@
+using System;
+using Microsoft.Data.SqlClient;
+
+/// The purpose of this sample is to illustrate how to use this feature and the condition might not be realistic.
```
Comment (johnnypham) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_OpenConnection.cs:4]:
```suggestion
// The purpose of this sample is to illustrate how to use this feature and the condition might not be realistic.
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_OpenConnection.cs [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+    }
+    private static void RetryConnection(string connectionString)
+    {
+        // <Snippet1>
+        // Define the retry logic parameters
+        var options = new SqlRetryLogicOption()
+        {
+            // Tries 5 times before throwing an exception
+            NumberOfTries = 5,
+            // Prefered gap time to delay before retry
```
Comment (johnnypham) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_OpenConnection.cs:21]:
```suggestion
            // Preferred gap time to delay before retry
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs (hunk 11 above)
Comment (DavoudEshtehari) [reply doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:12]: It's not included in Snippet blocks. But, I'd rather to remove it. 👍

**Review by karinazhou — APPROVED** [review]

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+        private static bool TransientErrorsCondition(Exception e, IEnumerable<int> retriableConditions)
+        {
+            bool result = false;
+
+            // Assess only SqlExceptions
+            if (retriableConditions != null && e is SqlException ex)
+            {
+                foreach (SqlError item in ex.Errors)
+                {
+                    // Check each error numbers if it was listed as a retriable error number
```
Comment (David-Engel) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:80]:
```suggestion
                    // Check each error number to see if it is a retriable error number
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+                    if (retriableConditions.Contains(item.Number))
+                    {
+                        result = true;
+                        break;
+                    }
+                }
+            }
+            // Other types of exceptions can be assessed separately
```
Comment (David-Engel) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:88]:
```suggestion
            // Other types of exceptions can also be assessed
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 22]
```diff
@@ -0,0 +1,254 @@
...
+        public CustomEnumerator(TimeSpan timeInterval, TimeSpan maxTime, TimeSpan minTime)
+            : base(timeInterval, maxTime, minTime) {}
+
+        // Return fixed time on each request
+        protected override TimeSpan GetNextInterval()
+        {
+            return GapTimeInterval;
+        }
+
+        // Override the validate method to apply the new time range validation
```
Comment (David-Engel) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:113]:
```suggestion
        // Override the validate method with the new time range validation
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 23]
```diff
@@ -0,0 +1,254 @@
...
+                throw new ArgumentOutOfRangeException(nameof(minTimeInterval));
+            }
+        }
+    }
+    // </Snippet6>
+
+    // <Snippet7>
+    public class CustomRetryLogic : SqlRetryLogicBase
+    {
+        // Maximum number of acceptable attempts
```
Comment (David-Engel) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:142]:
```suggestion
        // Maximum number of attempts
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 24]
```diff
@@ -0,0 +1,254 @@
...
+                                 SqlRetryIntervalBaseEnumerator enumerator,
+                                 Predicate<Exception> transientPredicate)
+        {
+            if (!(numberOfTries > 0 && numberOfTries <= maxAttempts))
+            {
+                // 'numberOfTries' should be between 1 and 20.
+                throw new ArgumentOutOfRangeException(nameof(numberOfTries));
+            }
+
+            // Assign the given parameters to the relevant properties
```
Comment (David-Engel) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:155]:
```suggestion
            // Assign parameters to the relevant properties
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 25]
```diff
@@ -0,0 +1,254 @@
...
+        {
+            Current = 0;
+            RetryIntervalEnumerator.Reset();
+        }
+
+        public override bool TryNextInterval(out TimeSpan intervalTime)
+        {
+            intervalTime = TimeSpan.Zero;
+            // First try has occurred before starting the retry process. 
+            // Check if it's still allowed to do retry
```
Comment (David-Engel) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:173]:
```suggestion
            // Check if retry is still allowed
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 26]
```diff
@@ -0,0 +1,254 @@
...
+            bool result = Current < NumberOfTries - 1;
+
+            if (result)
+            {
+                // Increase the number of attempts
+                Current++;
+                // It doesn't matter if the enumerator gets to the last value until the number of attempts end.
```
Comment (David-Engel) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:180]: This statement is confusing. What is it trying to say?

I'm guessing:
??? "It's okay if the enumerator doesn't get to the last value by the time we've reach our maximum number of attempts." ???

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 27]
```diff
@@ -0,0 +1,254 @@
...
+            RetryLogic.Reset();
+            // Create an infinite loop to attempt the defined maximum number of tries
+            do
+            {
+                try
+                {
+                    // Try to invoke the function
+                    return function.Invoke();
+                }
+                // Catch any type of exceptions for further investigations
```
Comment (David-Engel) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:213]:
```suggestion
                // Catch any type of exception for further investigation
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 28]
```diff
@@ -0,0 +1,254 @@
...
+                catch (Exception e)
+                {
+                    // Ask the RetryLogic object if this exception is a transient error
+                    if (RetryLogic.TransientPredicate(e))
+                    {
+                        // Add the exception to the list of retriable exceptions
```
Comment (David-Engel) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:219]:
```suggestion
                        // Add the exception to the list of exceptions we've retried on
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 29]
```diff
@@ -0,0 +1,254 @@
...
+                        exceptions.Add(e);
+                        // Ask the RetryLogic for the next delay time before next attempt to run the function
```
Comment (David-Engel) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:221]:
```suggestion
                        // Ask the RetryLogic for the next delay time before the next attempt to run the function
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs [hunk 30]
```diff
@@ -0,0 +1,254 @@
...
+                        if (RetryLogic.TryNextInterval(out TimeSpan gapTime))
+                        {
+                            Console.WriteLine($"Wait for {gapTime} before next try");
+                            // Wait before next attempt
+                            Thread.Sleep(gapTime);
+                        }
+                        else
+                        {
+                            // Number of attempts has exceeded from the maximum number of tries
```
Comment (David-Engel) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:230]:
```suggestion
                            // Number of attempts has exceeded the maximum number of tries
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_OpenConnection.cs [hunk 31]
```diff
@@ -0,0 +1,44 @@
+using System;
+using Microsoft.Data.SqlClient;
+
+// The purpose of this sample is to illustrate how to use this feature and the condition might not be realistic.
```
Comment (David-Engel) [inline doc/samples/SqlConfigurableRetryLogic_StepByStep_OpenConnection.cs:4]:
```suggestion
// The purpose of this sample is to illustrate how to use this feature and the example conditions might not be realistic.
```

File: doc/snippets/Microsoft.Data.SqlClient/SqlCommand.xml [hunk 32]
```diff
@@ -2796,12 +2796,15 @@ To apply the retry logic, do the following steps before executing the command:
...
-> [NOTE!]
+> [!NOTE]
 > Detecting retriable exceptions is a vital part of the retry pattern. Before applying retry logic, it is important to investigate exceptions and choose a retry provider that best fits your scenario. First, log your exceptions and find transient faults.  
 
-> [NOTE!]
+> [!NOTE]
 > The command **timeout** restarts for each execution of a command within the retry logic and after applying the retry time delay. There is no timing overlap between these two actions.
 
+> [!NOTE]
+> The default retry logic provider is not enabled unless it was configured in application configuration file. For more information, see [Configurable retry logic and configuration file](/sql/connect/ado-net/configurable-retry-logic-config-file).
```
Comment (David-Engel) [inline doc/snippets/Microsoft.Data.SqlClient/SqlCommand.xml:2806]:
```suggestion
> The default retry logic provider is not enabled unless it is configured in an application configuration file. For more information, see [Configurable retry logic configuration file](/sql/connect/ado-net/configurable-retry-logic-config-file-sqlclient).
```

File: doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml [hunk 33]
```diff
@@ -1058,12 +1058,15 @@ To apply the retry logic, do the following steps before opening the connection:
...
-> [NOTE!]
+> [!NOTE]
 > Detecting retriable exceptions is a vital part of the retry pattern. Before applying retry logic, it is important to investigate exceptions and choose a retry provider that best fits your scenario. First, log your exceptions and find transient faults.  
 
-> [NOTE!]
+> [!NOTE]
 > The connection **timeout** restarts for each execution of a connection open. There is no timing overlap between these two actions.
 
+> [!NOTE]
+> The default retry logic provider is not enabled unless it was configured in application configuration file. For more information, see [Configurable retry logic and configuration file](/sql/connect/ado-net/configurable-retry-logic-config-file).
```
Comment (David-Engel) [inline doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml:1068]:
```suggestion
> The default retry logic provider is not enabled unless it is configured in an application configuration file. For more information, see [Configurable retry logic and configuration file](/sql/connect/ado-net/configurable-retry-logic-config-file-sqlclient).
```

File: doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs (hunk 26 above)
Comment (DavoudEshtehari) [reply doc/samples/SqlConfigurableRetryLogic_StepByStep_CustomProvider.cs:180]: `MoveNext()` returns `false` when it reached the last sequence of interval-times. Here I've just assessed the maximum number of attempts, and it doesn't matter if it is repeating a delay-time, multiple times.
I hope my explanation made it a bit more clear.

**Pull request merged**